### PR TITLE
Branch putio client

### PR DESCRIPTION
**Pull request by devclinton**
Fixes # Feature Request  http://feathub.com/SickRage/SickRage/+27

Proposed changes in this pull request:
- Add support for put.io. Can be paired with http://oudenmallon.github.io/putioDownloader/ to allow full cycle from downloading to then sorting using SickRage post processing

### CODE REVIEW

**Comment by SickRage2** [comment]
Thanks for the pull request! Before a real human comes by, please make sure your PR has all of the below criteria checked
- [x] Give a description on what the PR is for.
- [ ] Make sure your PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review

Please make sure you also read [contribution guide](https://github.com/SickRage/SickRage/blob/master/contributing.md) and followed all of the steps.

Thanks!

**Comment by miigotu** [comment]
Few things here, first you need to PR to develop, not master. Second, vendor.min.js should not have changed. Did you test this? Is it working perfectly?

I will close this so you can make the PR to develop, thanks

File: sickbeard/clients/putio_client.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import sickbeard
+from sickbeard import logger
+from .generic import GenericClient
+
+class PutioAPI(GenericClient):
+    def __init__(self, host=None, username=None, password=None):
+
+        super(PutioAPI, self).__init__('put_io', host, username, password)
+
+        self.url = 'https://api.put.io/v2/oauth2/authenticate?client_id=2103&response_type=token&redirect_uri=http%3A%2F%2Fdevclinton.com'
```
Comment (OmgImAlexis) [inline sickbeard/clients/putio_client.py:31]: This also shouldn't be used. Users should be redirected to sickrage's github page if a default placeholder is needed and not your own website.

**Comment by devclinton** [comment]
My apologies. I will redo on development branch and submit again tomorrow
after work. Thanks for your patience!

File: sickbeard/clients/putio_client.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    def _get_auth(self):
+
+        post_data = {
+            'name': self.username,
+            'password': self.password,
+            'next': '/v2/oauth2/authenticate?client_id=2103&response_type=token&redirect_uri=http://devclinton.com'
```
Comment (OmgImAlexis) [inline sickbeard/clients/putio_client.py:38]: Same for here. @miigotu you should get a new client key from the account you setup so it's an "official" one and not just one that the person making this PR has as it'll belong to their account.

File: sickbeard/clients/putio_client.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            self.response = self.session.get(self.url, timeout=120)
+            logger.log("Getting authenicate page: %s" % self.response, logger.INFO)
+            self.tok_response = self.session.post('https://api.put.io/login',
+                data=post_data, allow_redirects=False)
+            logger.log("PutIO Authencate Form: %s. Headers: %s " % ( self.tok_response, self.tok_response.headers )  , logger.INFO )
+            logger.log("PutIO Authencate Form Body: \n%s " % ( self.tok_response.text )  , logger.INFO )
+            logger.log("PutIO Url: %s" % self.tok_response.headers['location'], logger.INFO)
+            self.tok_response = self.session.get(self.tok_response.headers['location'], allow_redirects=False)
+            logger.log("PutIO Authencate Response: %s. Headers: %s " % ( self.tok_response, self.tok_response.headers )  , logger.INFO )
+            self.auth = re.search('http://devclinton\.com#access_token=(.*)', self.tok_response.headers['location']).group(1)
```
Comment (OmgImAlexis) [inline sickbeard/clients/putio_client.py:53]: Same as here.

File: sickbeard/clients/putio_client.py (hunk 1 above)
Comment (miigotu) [inline sickbeard/clients/putio_client.py:31]: I agree (although I dont think that is his site, its the one in the cp client I think)

Comment (OmgImAlexis) [inline sickbeard/clients/putio_client.py:31]: Nope, it's @oudenmallon https://keybase.io/devclinton

File: sickbeard/clients/putio_client.py (hunk 3 above)
Comment (OmgImAlexis) [inline sickbeard/clients/putio_client.py:53]: We should be able to just create a page on the sickrage site that gets the URL using js and adds a div with the access token.

**Comment by miigotu** [comment]
Change out devclinton.com for https://sickrage.github.io and the client id is 2392

File: sickbeard/clients/putio_client.py (hunk 3 above)
Comment (miigotu) [inline sickbeard/clients/putio_client.py:53]: In this client the user is never redirected anywhere, it just gets the code from the redirect headers in requests.

So, the site isnt really important, but nevertheless https://github.com/SickRage/SickRage/pull/1468#issuecomment-209221785

File: sickbeard/clients/putio_client.py (hunk 1 above)
Comment (miigotu) [inline sickbeard/clients/putio_client.py:31]: @oudenmallon can I haz invite to keybase? miigotu at gmail if so xDD

**Comment by miigotu** [comment]
I guess I should take this over? =P  Let me know if you still plan to do this, thanks.

File: sickbeard/clients/putio_client.py (hunk 1 above)
Comment (devclinton) [inline sickbeard/clients/putio_client.py:31]: Yes, this is the put.io client register link. My updated pull request will have more comments around this, but when you register with the put.io API, you must register a return URL for user approvals. In this implementation though, I grab the token from the Redirect Address but don't actually redirect the user.

Comment (miigotu) [inline sickbeard/clients/putio_client.py:31]: In my other PR, I added the sickrage-ng application which has the url as https://sickrage.github.io

**Pull request closed without merging**